### PR TITLE
chore(sql): more precise version of regr_slope

### PR DESCRIPTION
**Pull request by nwoolmer**
There was some tooing and frooing about how to calculate this value. From testing, this version adds +3 d.p on trades dataset, and on artificial data.

On an artificial dataset (100 million random doubles), and on first 1 million trades from demo:

```sql
select covar_pop(b, a) / var_pop(a) from bench;
-- 230 ms accurate           to 0.000166195534833.....

select regr_slope(b, a) from bench;
-- 170 ms accurate           to 0.000166195534833.....
-- old version only accurate to 0.000166195534........

select covar_pop(price, amount) / var_pop(amount) from trades where symbol = 'ETH-USD'
-- 2-4ms accurate            to 2.03333734899.

select regr_slope(price, amount) from trades where symbol = 'ETH-USD'
-- 2-4ms accurate            to 2.03333734899.
-- old version only accurate to 2.03333734....
```

Its worth having the function on its own and not just wrapping the other two functions in a division, as it can make a 20-25% difference on larger datasets. Plus doesn't work with nulls, since var will still calculate rows that have a null in the other column.

## Why didn't it work in the original PR?

Fyi @vss96 !

The merging logic for the sums wasn't quite right. Since the sums were not plain summations, but were actually sums of squares, there's some extra logic needed. Original logic did this:

```java
mergedSumXY = srcSumXY + destSumXY;
```
New logic does this:

```java
long mergedCount = srcCount + destCount;
double deltaY = destMeanY - srcMeanY;
double deltaX = destMeanX - srcMeanX;
double mergedSumXY = srcSumXY + destSumXY + (deltaY * deltaX) * ((double) (srcCount * destCount) / mergedCount);
double mergedSumX = srcSumX + destSumX + (deltaX * deltaX) * ((double) (srcCount * destCount) / mergedCount);
```

Which is what happens in the covar/var functions.

https://github.com/questdb/questdb/pull/5021/commits/46ce1e2bc3516499fd13d423565c83e32f856a7d#diff-b310ff8a4236fd3ea371cf7c4493e08c33521ae0aa9fc8131eae2e53d4ff5b25R207-R208

### CODE REVIEW

**Comment by vss96** [comment]
@nwoolmer I had this in my previous version of the [PR](https://github.com/questdb/questdb/pull/5021/commits/10dfab7ccdf50180418791a24fdb6ad9c2bfbeef) - we were having issues with numerical stability. You can read more about it in the comments but the problem with this implementation is that the values deviate as you recompute them in parallel.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 55 / 55 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/RegressionSlopeFunctionFactory.java|55|55|100.00%|

**Comment by vss96** [comment]
> @vss96 How much deviation was there? At least with the version in this PR, its 3 d.p more stable than the version we ended up merging (textbook version). Perhaps there was a different issue!

Previously we were getting  < 3 d.p - can't remember for sure.
I tried building up a codespace and checking this change out and it looks pretty stable. Now I'm not sure why the previous version wasn't stable 🤯

**Comment by vss96** [comment]
Had a quick look and it might have been the error compensation values that might have caused issues before - the values tend to be stable without them.

Thank you for looking into it and fixing it!

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 55 / 55 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/RegressionSlopeFunctionFactory.java|55|55|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 55 / 55 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/RegressionSlopeFunctionFactory.java|55|55|100.00%|

File: core/src/main/java/io/questdb/griffin/engine/functions/groupby/RegressionSlopeFunctionFactory.java [hunk 1]
```diff
@@ -154,23 +152,37 @@ public boolean isThreadSafe() {
...
+            double deltaX = destMeanX - srcMeanX;
+
+            // This is only valid when countA is much larger than countB.
+            // If both are large and similar sizes, delta is not scaled down.
+            // double mergedMean = srcMean + delta * ((double) destCount / mergedCount);
+
+            // So we use this instead:
+            double mergedMeanY = (srcCount * srcMeanY + destCount * destMeanY) / mergedCount;
+            double mergedMeanX = (srcCount * srcMeanX + destCount * destMeanX) / mergedCount;
+            double mergedSumXY = srcSumXY + destSumXY + (deltaY * deltaX) * ((double) (srcCount * destCount) / mergedCount);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/functions/groupby/RegressionSlopeFunctionFactory.java:178]: the long multiplication will overflow with incorrect value if you merge past 10Bn and 1Bn values. The `srcCount` should be cast to `double` before the multiplication.

Additionally, the cluster 

```java
(deltaY * deltaX) * ((double) (srcCount * destCount) / mergedCount)
```

can be pre-computed and reused in both `mergedSumXY` and `mergedSumX`. I would not trust Java to optimise that.

Comment (bluestreak01) [reply core/src/main/java/io/questdb/griffin/engine/functions/groupby/RegressionSlopeFunctionFactory.java:178]: ignore the extraction bit, i noticed that those are not the same calculations

Comment (bluestreak01) [reply core/src/main/java/io/questdb/griffin/engine/functions/groupby/RegressionSlopeFunctionFactory.java:178]: I would extract:

```java
double x = (((double) srcCount * destCount) / mergedCount)
```

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 58 / 58 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/CovarSampleGroupByFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/RegressionSlopeFunctionFactory.java|56|56|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/VarSampleGroupByFunctionFactory.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 58 / 58 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/CovarSampleGroupByFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/RegressionSlopeFunctionFactory.java|56|56|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/VarSampleGroupByFunctionFactory.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 58 / 58 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/CovarSampleGroupByFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/RegressionSlopeFunctionFactory.java|56|56|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/VarSampleGroupByFunctionFactory.java|1|1|100.00%|

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**